src/core/account.ts

```diff
@@ -97,62 +99,77 @@ export class Account {
-   * should be used only with private key that was created
-   * as
-   *
-   * @param privateKey Hex - private key of the account
-   * @returns Account
-   */
-  static fromLegacyPrivateKey(privateKey: PrivateKey) {
-    const publicKey = privateKey.publicKey();
-    const authKey = Account.authKey({ publicKey });
-    const address = new AccountAddress({ data: authKey.toUint8Array() });
-    return Account.fromLegacyPrivateKeyAndAddress({ privateKey, address });
-  }
+    if (privateKey instanceof Secp256k1PrivateKey) {
+      // private key is secp256k1, therefore we know it for sure uses a single signer key
+      const authKey = AuthenticationKey.fromBytesAndScheme({ publicKey, scheme: 2 });
+      const address = new AccountAddress({ data: authKey.toUint8Array() });
+      return new Account({ privateKey, address });
+    }
 
-  /**
-   * Derives an account with provided private key and address
-   * This is intended to be used for account that has it's key rotated
-   *
-   * @param args.privateKey Hex - private key of the account
-   * @param args.address AccountAddress - address of the account
-   * @returns Account
-   */
-  static fromPrivateKeyAndAddress(args: { privateKey: PrivateKey; address: AccountAddress }): Account {
-    return new Account(args);
+    if (privateKey instanceof Ed25519PrivateKey) {
+      // lookup single sender ed25519
+      const singleSenderAuthKey = AuthenticationKey.fromBytesAndScheme({ publicKey, scheme: SigningScheme.SingleKey });
+      const isSingleSender = await Account.lookupAddress(singleSenderAuthKey, config);
+      if (isSingleSender) {
+        const address = new AccountAddress({ data: singleSenderAuthKey.toUint8Array() });
+        return new Account({ privateKey, address });
+      }
+      // lookup legacy ed25519
+      const legacyAuthKey = AuthenticationKey.fromBytesAndScheme({ publicKey, scheme: SigningScheme.Ed25519 });
+      const isLegacyEd25519 = await Account.lookupAddress(legacyAuthKey, config);
+      if (isLegacyEd25519) {
+        const address = new AccountAddress({ data: legacyAuthKey.toUint8Array() });
+        return new Account({ privateKey, address, legacy: true });
+      }
+    }
+
+    // if we are here, it means we couldn't find an address with an
+    //auth key that matches the provided private key
+    throw new Error(`Can't derive account from private key ${privateKey}`);
   }
 
-  static fromLegacyPrivateKeyAndAddress(args: { privateKey: PrivateKey; address: AccountAddress }): Account {
-    return new Account(args, true);
+  private static async lookupAddress(authKey: AuthenticationKey, config: AptosConfig): Promise<boolean> {
+    const potentialAddress1 = await lookupOriginalAccountAddress({
+      aptosConfig: config,
+      authenticationKey: authKey.toString(),
+    });
+
+    try {
+      await getInfo({
+        aptosConfig: config,
+        accountAddress: potentialAddress1.toString(),
+      });
+      return true;
+    } catch (error: any) {
```

Comment:
I think we should only swallow the error if it's `resource_not_found` or whatever, that way if it's some other API error, we're not assuming the account wasn't found and throwing an actual error